src/server/engine_shard_set.cc

```diff
@@ -545,29 +540,36 @@ void EngineShard::PollExecution(const char* context, Transaction* trans) {
     DVLOG(1) << "Skipped TxQueue " << continuation_trans_;
   }
 
-  // we need to run trans if it's OOO or when trans is blocked in this shard.
-  bool should_run = trans_mask & (Transaction::OUT_OF_ORDER | Transaction::SUSPENDED_Q);
+  // If the caller isn't armed anymore, execution was handled in the txq loop by an earlier poll.
+  // Otherwise, we determine if we can run the caller transaction now.
+  if (!trans || !trans->IsArmedInShard(sid))
+    return;
+
+  auto trans_mask = trans->GetLocalMask(sid);
 
-  // It may be that there are other transactions that touch those keys but they necessary ordered
-  // after trans in the queue, hence it's safe to run trans out of order.
-  if (trans && should_run) {
+  // OOO means no transaction before us in the txq accesses our keys, so we can run earlier
+  bool is_ooo = trans_mask & Transaction::OUT_OF_ORDER;
+
+  // Still suspended shards need to run just to finalize and unregister, so we can run anytime
+  bool is_suspended = trans_mask & Transaction::SUSPENDED_Q;
+  DCHECK_EQ(trans_mask & Transaction::AWAKED_Q, 0);
+
```

Comment:
This is a little bit confusing. When a blocking tx unblocks, only one shard wakes up, all others keep "suspended". But all shards need to run, so the suspended ones can clean up. They actually call the callback as well, it's just a no-op (should be) for them. 

It wasn't documented before and I was really confused for half an hour why we'd run suspended transactions unconditionally. The key is: they coordinator is already woken up and they need to finalize